crates/matrix-sdk-ui/src/timeline/controller/decryption_retry_task.rs

```diff
@@ -119,16 +121,33 @@ async fn decryption_task<D: Decryptor>(
             }
         };
 
-        let retry_indices = item_indices_to_retry(state.clone(), &should_retry).await;
-        if !retry_indices.is_empty() {
+        // Find the indices of events that are in the supplied sessions, distinguishing
+        // between UTDs which we need to decrypt, and already-decrypted events where we
+        // only need to re-fetch encryption info.
+        let (retry_decryption_indices, retry_info_indices) =
+            event_indices_to_retry_decryption(state.read().await.items.iter(), should_retry);
```

Comment:
Can you get the state lock once, instead of getting it once in read mode here, and twice in write mode below? Otherwise it could be taken elsewhere, and the items would be modified in the meanwhile, possibly causing super weird confusion for the indices…